### PR TITLE
serde compatibility layer

### PR DESCRIPTION
**Pull request by mitjat**
Resolves https://oasisprotocol.clickup.com/t/27xhuvj

### CODE REVIEW

File: src/decode.rs [hunk 1]
```diff
@@ -165,6 +165,21 @@ impl Decode for String {
     }
 }
 
+impl Decode for char {
+    fn try_default() -> Result<Self, DecodeError> {
+        Ok(Default::default())
```
Comment (kostko) [inline src/decode.rs:170]: Add test case to `test_null_decode`.

File: src/encode.rs [hunk 2]
```diff
@@ -155,6 +155,12 @@ impl Encode for &str {
     }
 }
 
+impl Encode for char {
```
Comment (kostko) [inline src/encode.rs:158]: Should probably add an `is_empty` implementation (e.g. if the numeric value is equal to zero) together with a test in `encode_empty`.

File: src/serde/de.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+use crate::Value;
+
+use super::error::{unexpected, Error};
+use crate::Decode as _;
+use serde::{
+    de::{self, Error as _, Visitor},
+    Deserializer as _,
+};
+use sk_cbor::SimpleValue;
```
Comment (kostko) [inline src/serde/de.rs:11]: Move this into a separate group above (e.g. third party imports).

File: src/serde/de.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
+use std::iter::Peekable;
+
+use crate::Value;
```
Comment (kostko) [inline src/serde/de.rs:3]: Move this below (e.g. group under "crate-internal imports").

File: src/serde/de.rs [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+            fn next_key_seed<K>(&mut self, seed: K) -> Result<Option<K::Value>, Self::Error>
+            where
+                K: de::DeserializeSeed<'de>,
+            {
+                match self.0.peek_mut() {
+                    Some((k, _v)) => {
+                        let k = {
+                            // Simulate k.take() in face of the fact that `Value` does not impl `Default`.
+                            let mut replacement = Value::Simple(SimpleValue::Undefined);
+                            std::mem::swap(k, &mut replacement);
```
Comment (kostko) [inline src/serde/de.rs:137]: Why not `let k = std::mem::replace(k, Value::Simple(SimpleValue::Undefined))`?

File: src/serde/error.rs [hunk 6]
```diff
@@ -0,0 +1,64 @@
+#![cfg(feature = "serde")]
```
Comment (kostko) [inline src/serde/error.rs:1]: Why is this needed?

File: src/serde/mod.rs [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+//! Although it provides convenience methods for (de)serializing directly into/from bytes,
+//! this module only implements converts between native rust types and `sk_cbor::Value`. The
+//! conversion between `sk_cbor::Value` and bytes is handled by the `sk_cbor` library.
+//!
+//! NOTE: CBOR encoding is not strictly defined for non-primitive types. For any given type T,
+//! serializations produced by `#[derive(Encode)]` (from core `oasis-cbor`) and those produced
+//! by `#[derive(Deserialize)]` (from this module) are NOT GUARANTEED TO BE COMPATIBLE.
+//! See notes in the test suite for examples of incompatibilities.
+
+#![cfg(feature = "serde")]
```
Comment (kostko) [inline src/serde/mod.rs:11]: Why is this needed given that parent already specifies it?

File: src/serde/ser.rs [hunk 8]
```diff
@@ -0,0 +1,328 @@
+use core::result::Result;
```
Comment (kostko) [inline src/serde/ser.rs:1]: Why is this needed, `Result` is part of the implicit prelude import.

File: src/serde/ser.rs [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+
+use serde::{
+    ser::{self, SerializeStruct},
+    Serialize,
+};
+
+use crate::{serde::error::Error, Encode, Value};
+
+pub(crate) struct Serializer {}
```
Comment (kostko) [inline src/serde/ser.rs:10]: You could define it as:
```suggestion
pub(crate) struct Serializer;
```
To make construction simpler.

File: src/serde/ser.rs [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    ) -> Result<(), Self::Error> {
+        let mut s = Serializer {};
+        self.fields
+            .push((key.serialize(&mut s)?, value.serialize(&mut s)?));
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        // Normalization (sorting) also happens inside sk_cbor, so this is
+        // technically redundant, but oasis_cbor (non-serde) does the same,
```
Comment (kostko) [inline src/serde/ser.rs:220]: The reason for this additional sorting is to be able to decode from intermediate `Value` types as otherwise `cbor::from_value` may not work (since sorting is mandatory in canonical CBOR) on values produced by `cbor::to_value`.

File: src/serde/ser.rs [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+    ) -> Result<(), Self::Error> {
+        self.fields.push((
+            key.serialize(&mut Serializer {})?,
+            value.serialize(&mut Serializer {})?,
+        ));
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        Ok(Value::Map(self.fields))
```
Comment (kostko) [inline src/serde/ser.rs:326]: No sorting?

File: src/serde/ser.rs [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+            self.key
+                .take()
+                .expect("serde tried to encode a map value without a key"),
+            value.serialize(&mut Serializer {})?,
+        ));
+        Ok(())
+    }
+
+    fn end(self) -> Result<Self::Ok, Self::Error> {
+        // Normalization (sorting) also happens inside sk_cbor, so this is
```
Comment (kostko) [inline src/serde/ser.rs:297]: Note that we currently only sort for struct (and variant struct) fields, not for regular maps.

File: src/serde/test.rs [hunk 13]
```diff
@@ -0,0 +1,325 @@
+#![cfg(test)]
```
Comment (kostko) [inline src/serde/test.rs:1]: This should be in the parent `mod.rs` instead.

File: src/serde/test.rs [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+
+use crate::{SimpleValue, Value};
+
+use serde::{Deserialize, Serialize};
```
Comment (kostko) [inline src/serde/test.rs:5]: Change import group order:
* Standard library.
* External crates.
* Internal imports.

File: src/serde/mod.rs [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+where
+    T: ?Sized + serde::Serialize,
+{
+    let mut data = vec![];
+    sk_cbor::writer::write(as_cbor_value(value)?, &mut data).map_err(Error::from)?;
+    Ok(data)
+}
+
+/// Serialize `value` into intermediate-representation `sk_cbor::Value`.
+pub fn as_cbor_value<T>(value: &T) -> Result<Value, Error>
```
Comment (kostko) [inline src/serde/mod.rs:51]:
```suggestion
pub fn to_value<T>(value: &T) -> Result<Value, Error>
```

File: src/serde/mod.rs [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+where
+    T: serde::de::DeserializeOwned,
+{
+    let value =
+        sk_cbor::reader::read_nested(data, Some(crate::MAX_NESTING_LEVEL)).map_err(Error::from)?;
+    from_cbor_value(value)
+}
+
+/// Deserialize intermediate-representation `sk_cbor::Value` into `T`.
+pub fn from_cbor_value<T>(value: Value) -> Result<T, Error>
```
Comment (kostko) [inline src/serde/mod.rs:32]:
```suggestion
pub fn from_value<T>(value: Value) -> Result<T, Error>
```

**Review by kostko — COMMENTED** [review]
Nice work!

Added some comments, also add (failing) tests for serializing/deserializing floats (serde and non-serde).

File: src/serde/ser.rs (hunk 11 above)
Comment (mitjat) [reply src/serde/ser.rs]: Good catch. It was bad luck that the roundtrip test (encode with serde, decode with plain oasis-cbor) didn't catch this; the "random" order of keys just happened to be sorted. Added sorting + explanation.

File: src/serde/ser.rs (hunk 12 above)
Comment (mitjat) [reply src/serde/ser.rs]: Removed sorting, added explanation.

File: src/serde/error.rs (hunk 6 above)
Comment (mitjat) [reply src/serde/error.rs]: My `cfg` were all over the place, mostly because I refactored once or twice as I went. Cleaned up now, here and elsewhere.

File: src/serde/ser.rs (hunk 8 above)
Comment (mitjat) [reply src/serde/ser.rs]: removed

**Review by mitjat — COMMENTED** [review]
> also add (failing) tests for serializing/deserializing floats (serde and non-serde).

Added a test in serde, although the `sk_cbor::Value` type with its absence of floats mostly already forces our hand there.
In non-serde code, `derive(Decode)` complains at compile time already because there's no `impl Decode for f32`. Doesn't seem worth adding a test for that.

File: src/serde/ser.rs (hunk 10 above)
Comment (kostko) [reply src/serde/ser.rs]: I agree with your approach, let's fix decode and get rid of the value sorts in encoders.

**Review by mitjat — COMMENTED** [review]
Thank you for finding the time, Jernej!
Have another look, I think we're probably good to go now.

**Comment by kostko** [comment]
It seems you forgot to push the last batch of changes :-)
